src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -634,21 +639,30 @@ def query(self, filters=None, options=None):
             )
             if cp.returncode != 0:
                 raise CallError(f'Failed to retrieve snapshots: {cp.stderr}')
-            snaps = [{'name': i} for i in cp.stdout.strip().split('\n')]
+            snaps = [
+                {'name': i, 'pool': i.split('/', 1)[0]}
+                for i in cp.stdout.strip().split('\n')
+            ]
             if filters:
                 return filter_list(snaps, filters, options)
             return snaps
         with libzfs.ZFS() as zfs:
             # Handle `id` filter to avoid getting all snapshots first
+            snapshots = []
             if filters and len(filters) == 1 and list(filters[0][:2]) == ['id', '=']:
                 try:
-                    snapshots = [zfs.get_snapshot(filters[0][2]).__getstate__()]
+                    snapshots.append(zfs.get_snapshot(filters[0][2]).__getstate__())
                 except libzfs.ZFSException as e:
                     if e.code != libzfs.Error.NOENT:
                         raise
-                    snapshots = []
             else:
-                snapshots = [i.__getstate__() for i in list(zfs.snapshots)]
+                for i in zfs.snapshots:
+                    try:
+                        snapshots.append(i.__getstate__())
```

Comment:
How about we retrieve a serialized version directly from py-libzfs ? It gives us better performance.
I'll double check to be sure but i think the format of getstate is respected thus avoiding any unintended consequences/surprises.

Reply:
This needs to be backported for 11.2, so not yet.